R/step_epi_slide.R

```diff
@@ -0,0 +1,277 @@
+           after = 0L,
+           role = "predictor",
+           prefix = "epi_slide_",
+           f_name = clean_f_name(.f),
+           skip = FALSE,
+           id = rand_id("epi_slide")) {
+    if (!is_epi_recipe(recipe)) {
+      rlang::abort("This recipe step can only operate on an `epi_recipe`.")
+    }
+    .f <- validate_slide_fun(.f)
+    arg_is_scalar(before, after)
+    before <- try_period(before)
+    after <- try_period(after)
+    arg_is_chr_scalar(role, prefix, id)
+    arg_is_lgl_scalar(skip)
+
+    add_step(
+      recipe,
+      step_epi_slide_new(
+        terms = enquos(...),
+        before = before,
+        after = after,
+        .f = .f,
+        f_name = f_name,
+        role = role,
+        trained = FALSE,
+        prefix = prefix,
+        keys = epi_keys(recipe),
+        columns = NULL,
+        skip = skip,
+        id = id
+      )
+    )
+  }
+
+
+step_epi_slide_new <-
+  function(terms,
+           before,
+           after,
+           .f,
+           f_name,
+           role,
+           trained,
+           prefix,
+           keys,
+           columns,
+           skip,
+           id) {
+    step(
+      subclass = "epi_slide",
+      terms = terms,
+      before = before,
+      after = after,
+      .f = .f,
+      f_name = f_name,
+      role = role,
+      trained = trained,
+      prefix = prefix,
+      keys = keys,
+      columns = columns,
+      skip = skip,
+      id = id
+    )
+  }
+
+
+
+#' @export
+prep.step_epi_slide <- function(x, training, info = NULL, ...) {
+  col_names <- recipes::recipes_eval_select(x$terms, data = training, info = info)
+
+  check_type(training[, col_names], types = c("double", "integer"))
+
+  step_epi_slide_new(
+    terms = x$terms,
+    before = x$before,
+    after = x$after,
+    .f = x$.f,
+    f_name = x$f_name,
+    role = x$role,
+    trained = TRUE,
+    prefix = x$prefix,
+    keys = x$keys,
+    columns = col_names,
+    skip = x$skip,
+    id = x$id
+  )
+}
+
+
+
+#' @export
+bake.step_epi_slide <- function(object, new_data, ...) {
+  recipes::check_new_data(names(object$columns), object, new_data)
+  col_names <- object$columns
+  name_prefix <- paste0(object$prefix, object$f_name, "_")
+  newnames <- glue::glue("{name_prefix}{col_names}")
+  ## ensure no name clashes
+  new_data_names <- colnames(new_data)
+  intersection <- new_data_names %in% newnames
+  if (any(intersection)) {
+    nms <- new_data_names[intersection]
+    cli_abort(
+      c("Name collision occured. The following variable names already exist:",
+        `*` = "{.var {nms}}"
+      ),
+      call = caller_env()
+    )
+  }
+
+  ok <- object$keys
+  names(col_names) <- newnames
+  gr <- new_data %>%
+    dplyr::select(dplyr::all_of(c(ok, object$columns))) %>%
+    group_by(dplyr::across(dplyr::all_of(ok[-1]))) %>%
+    dplyr::arrange(time_value) %>%
+    dplyr::mutate(
+      dplyr::across(
+        dplyr::all_of(object$columns),
+        ~ slider::slide_index_vec(
```

Comment:
I had hoped to just use `epi_slide()`, but the interface for functions and column selections is very different. I'm happy to take suggestions, but `step_*()` need to have named args (in this case the function) and columns to apply the function to in `...`. I couldn't figure out how to convert that to `epi_slide()` syntax.

Reply:
so messing around a bit, I think we need to add a way of specifying columns to the generic `epi_slide`. if `.f=frollmean`, then we can use `epi_slide_opt` in a natural way:
```R
  col_names <- paste("slide_value", col_names, sep = "_")
  names(col_names) <- newnames
  new_data %>%
    dplyr::select(dplyr::all_of(c(ok, object$columns))) %>%
    group_by(dplyr::across(dplyr::all_of(ok[-1]))) %>%
    epiprocess::epi_slide_opt(
      !!col_name_orig,
      object$.fb,
      before = object$before,
      after = object$after) %>%
    rename(col_names)
```

Reply:
@nmdefries @lcbrooks

Reply:
(Potential solutions [down below](https://github.com/cmu-delphi/epipredict/pull/364#issuecomment-2251322343).  Probably should test & benchmark against more realistic data.  Though I'd suggest only allowing something backed by `epi_slide_opt` because of the forgetting-epikey-grouping, incomplete-window (gaps, initial values), and speed issues of `epi_slide`.)

Reply:
I did a bit of a better benchmark on slides1 and slides2, using `jhu_csse_daily_subset`, summing cases, and indeed, things do change.  Would still prefer `_opt`'s output and speed.

```
  expression                            min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result   memory time       gc      
  <bch:expr>                       <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>   <list> <list>     <list>  
1 slides1(edf, before, varnames, …    6.49s    6.49s     0.154        NA     5.09     1    33      6.49s <epi_df> <NULL> <bench_tm> <tibble>
2 slides2(edf, before, varnames, … 178.48ms 179.61ms     5.31         NA     5.31     3     3   564.77ms <epi_df> <NULL> <bench_tm> <tibble>
Warning message:
Some expressions had a GC in every iteration; so filtering is disabled. 
```